ompi/mca/coll/tuned/coll_tuned_dynamic_rules.c

```diff
@@ -357,12 +383,18 @@ int ompi_coll_tuned_get_target_method_params (ompi_coll_com_rule_t* base_com_rul
     /* make a copy of the first msg rule */
     best_msg_p = msg_p = base_com_rule->msg_rules;
     i = 0;
+    bool found_rules = false;
 
     while (i<base_com_rule->n_msg_sizes) {
         /*       OPAL_OUTPUT((ompi_coll_tuned_stream,"checking mpi_msgsize %d against com_id %d msg_id %d index %d msg_size %d",  */
         /*             mpi_msgsize, msg_p->com_rule_id, msg_p->msg_rule_id, i, msg_p->msg_size)); */
         if (msg_p->msg_size <= mpi_msgsize) {
-            best_msg_p = msg_p;
+            if (msg_p->topologic_level == DEFAULT ||
+               (msg_p->topologic_level == INTER_NODE && OMPI_COMM_IS_DISJOINT(comm)) || 
+               (msg_p->topologic_level == INTRA_NODE && !OMPI_COMM_IS_DISJOINT(comm))) {
```

Comment:
These 2 lines are wrong:

1. Disjoint communicator must be an intra-communicator, and the disjoint flag is set, so we should check for `OMPI_COMM_IS_INTRA && OMPI_COMM_IS_DISJOINT_SET && OMPI_COMM_IS_DISJOINT`
2. We want to check if the communicator is singlenode - it is totally different from `!OMPI_COMM_IS_DISJOINT`. I suggest reading #9934 to figure out a proper check.